<a href="https://colab.research.google.com/github/Ishan-malinda/Machine-Learning-Algorithm-for-Bank-Loan-System-Prediction/blob/main/Python_Notebook1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Notebook 1 — Data Understanding and Preprocessing
Author: W.F.ISHAN MALINDA SOVIS  


Purpose: Import the pandas library for data loading and manipulation

In [ ]:
# Import pandas library for data loading and manipulation
import pandas as pd

In [ ]:
# Set maximum rows pandas will display to 500
pd.set_option('display.max_rows', 500)
# Set maximum columns pandas will display to 500
pd.set_option('display.max_columns', 500)
# Set display width to 150 to prevent line wrapping
pd.set_option('display.width', 150)

Purpose: Load the coursework loan approval dataset from CSV file

In [ ]:
# Load the loan approval dataset from the uploaded CSV file
data = pd.read_csv('/content/drive/MyDrive/ML & DM/CW/loan_approval_data.csv')

Purpose: Display first 10 client records to inspect recorded data values per variable

In [ ]:
# Display the first 10 client records to inspect variable values
data.head(10)

,id,age,income,home_ownership,emplyment_length,loan_intent,loan_amount,loan_interest_rate,loan_income_ratio,payment_default_on_file,credit_history_length,loan_approval_status,max_allowed_loan
0,35437,21.0,12000,OWN,0,EDUCATION,15000,6.99,0.12,N,4,0,-2426900
1,53756,21.0,13200,OWN,2,EDUCATION,25000,16.77,0.19,Y,3,0,-111739
2,42205,23.0,9600,RENT,5,MEDICAL,30000,12.42,0.31,N,3,0,-89000
3,19180,40.0,182004,RENT,3,EDUCATION,35000,8.00,0.19,N,11,0,35000
4,28072,40.0,90000,MORTGAGE,3,HOMEIMPROVEMENT,35000,12.42,0.39,N,14,0,35000
5,32074,40.0,131004,MORTGAGE,9,VENTURE,30000,7.90,0.23,N,13,0,30000
6,7224,40.0,150000,MORTGAGE,9,DEBTCONSOLIDATION,25000,11.89,0.17,N,11,0,25000
7,14471,40.0,132000,MORTGAGE,3,HOMEIMPROVEMENT,25000,16.82,0.22,N,17,0,25000
8,20163,40.0,80000,MORTGAGE,2,DEBTCONSOLIDATION,25000,10.25,0.31,N,11,0,25000
9,20384,40.0,80000,MORTGAGE,9,VENTURE,25000,18.67,0.30,N,17,0,25000


Purpose: Display only the variable names from the dataset

In [ ]:
# Display a list of all variable names in the dataset
list(data.columns)

['id',
 'age',
 'income',
 'home_ownership',
 'emplyment_length',
 'loan_intent',
 'loan_amount',
 'loan_interest_rate',
 'loan_income_ratio',
 'payment_default_on_file',
 'credit_history_length',
 'loan_approval_status',
 'max_allowed_loan']

Purpose: Display data types and non-null counts for all variables to detect type issues

In [ ]:
# Display data types and non-null count for every variable in the dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58645 entries, 0 to 58644
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       58645 non-null  int64  
 1   age                      58639 non-null  float64
 2   income                   58645 non-null  int64  
 3   home_ownership           58645 non-null  object 
 4   emplyment_length         58645 non-null  int64  
 5   loan_intent              58645 non-null  object 
 6   loan_amount              58645 non-null  int64  
 7   loan_interest_rate       58634 non-null  float64
 8   loan_income_ratio        58645 non-null  float64
 9   payment_default_on_file  58640 non-null  object 
 10  credit_history_length    58645 non-null  int64  
 11  loan_approval_status     58645 non-null  int64  
 12  max_allowed_loan         58645 non-null  int64  
dtypes: float64(3), int64(7), object(3)
memory usage: 5.8+ MB


Purpose: Display the number of instances and features in the dataset

In [ ]:
# Display the exact number of rows (instances) and columns (features)
data.shape

(58645, 13)

Purpose: Calculate basic descriptive statistics for numeric variables to investigate data issues

In [ ]:
# Calculate and display descriptive statistics for all numeric variables
data.describe()

,id,age,income,emplyment_length,loan_amount,loan_interest_rate,loan_income_ratio,credit_history_length,loan_approval_status,max_allowed_loan
count,58645.000000,58639.000000,5.864500e+04,58645.000000,58645.000000,58634.000000,58645.000000,58645.000000,58645.000000,5.864500e+04
mean,29322.000000,27.550913,6.404617e+04,4.703487,9217.556518,10.677526,0.159238,5.813556,0.142382,6.975472e+04
std,16929.497605,6.033217,3.793111e+04,4.004982,5563.807384,3.036034,0.091692,4.029196,0.349445,6.175091e+04
min,0.000000,20.000000,4.200000e+03,0.000000,500.000000,-11.140000,0.000000,2.000000,0.000000,-2.426900e+06
25%,14661.000000,23.000000,4.200000e+04,2.000000,5000.000000,7.880000,0.090000,3.000000,0.000000,3.800300e+04
50%,29322.000000,26.000000,5.800000e+04,4.000000,8000.000000,10.750000,0.140000,4.000000,0.000000,6.239200e+04
75%,43983.000000,30.000000,7.560000e+04,7.000000,12000.000000,12.990000,0.210000,8.000000,0.000000,9.271600e+04
max,58644.000000,123.000000,1.900000e+06,150.000000,35000.000000,23.220000,0.830000,30.000000,1.000000,2.638778e+06


Purpose: Calculate descriptive statistics for categorical variables to inspect unique values

In [ ]:
# Calculate and display descriptive statistics for all categorical variables
data.describe(include='object')

,home_ownership,loan_intent,payment_default_on_file
count,58645,58645,58640
unique,4,6,2
top,RENT,EDUCATION,N
freq,30594,12271,49940


Purpose: Display distinct values for categorical variables to detect inconsistencies

In [ ]:
# Display unique values in home_ownership variable
print("home_ownership:", data['home_ownership'].unique())
# Display unique values in loan_intent variable
print("loan_intent:", data['loan_intent'].unique())
# Display unique values in payment_default_on_file variable
print("payment_default_on_file:", data['payment_default_on_file'].unique())
# Display unique values in loan_approval_status variable
print("loan_approval_status:", data['loan_approval_status'].unique())

home_ownership: ['OWN' 'RENT' 'MORTGAGE' 'OTHER']
loan_intent: ['EDUCATION' 'MEDICAL' 'HOMEIMPROVEMENT' 'VENTURE' 'DEBTCONSOLIDATION'
 'PERSONAL']
payment_default_on_file: ['N' 'Y' nan]
loan_approval_status: [0 1]


Purpose: Count missing values per variable to identify which variables have missing data

In [ ]:
# Count and display the number of missing values for each variable
data.isnull().sum()

,0
id,0
age,6
income,0
home_ownership,0
emplyment_length,0
loan_intent,0
loan_amount,0
loan_interest_rate,11
loan_income_ratio,0
payment_default_on_file,5


Purpose: Calculate percentage of missing values to assess severity and choose treatment method

In [ ]:
# Calculate the percentage of missing values per variable
data.isna().sum()/len(data)*100

,0
id,0.000000
age,0.010231
income,0.000000
home_ownership,0.000000
emplyment_length,0.000000
loan_intent,0.000000
loan_amount,0.000000
loan_interest_rate,0.018757
loan_income_ratio,0.000000
payment_default_on_file,0.008526


Purpose: Import plotly express package to visualise variables

In [ ]:
# Import plotly express for interactive data visualisation
import plotly.express as px

Purpose: Visualise frequency distribution of age variable to detect outliers

In [ ]:
# Plot histogram to show frequency distribution of the age variable
age_fig = px.histogram(data, x='age', title='Distribution of Age')
# Display the histogram
age_fig.show()

Purpose: Visualise frequency distribution of the target variable loan_approval_status

In [ ]:
# Count frequency of each loan approval status value
target_counts = data['loan_approval_status'].value_counts().reset_index()
# Plot bar chart showing distribution of target variable
target_fig = px.bar(target_counts, x='loan_approval_status', y='count',
                    title='Distribution of Loan Approval Status (0=Rejected, 1=Approved)')
# Display the chart
target_fig.show()

Purpose: Remove the id variable as it is a unique client identifier with no predictive value

In [ ]:
# Drop the id column as it is a unique identifier with no predictive value for modelling
data = data.drop(columns=(['id']))
# Confirm removal by listing remaining columns
list(data.columns)

['age',
 'income',
 'home_ownership',
 'emplyment_length',
 'loan_intent',
 'loan_amount',
 'loan_interest_rate',
 'loan_income_ratio',
 'payment_default_on_file',
 'credit_history_length',
 'loan_approval_status',
 'max_allowed_loan']

Purpose: Build a reusable function to detect outliers using the Inter-Quartile Range test

In [ ]:
# Define a function to detect outliers in any numeric variable using IQR method
def find_outliers_IQR(data):
    # Calculate the first quartile (25th percentile)
    q1 = data.quantile(0.25)
    # Calculate the third quartile (75th percentile)
    q3 = data.quantile(0.75)
    # Calculate the interquartile range
    IQR = q3 - q1
    # Identify values outside lower fence (Q1-1.5*IQR) and upper fence (Q3+1.5*IQR)
    outliers = data[((data < (q1 - 1.5*IQR)) | (data > (q3 + 1.5*IQR)))]
    return outliers

Purpose: Call the IQR function to detect outlier instances in the age variable

In [ ]:
# Call the IQR outlier detection function on the age variable
outliers = find_outliers_IQR(data['age'])
# Print the total count of detected outliers in age
print("number of outliers: " + str(len(outliers)))
# Display the outlier records
outliers

number of outliers: 2446


,age
8334,41.0
8335,41.0
8336,41.0
8337,41.0
8338,41.0
...,...
58602,43.0
58608,42.0
58613,55.0
58614,42.0


Purpose: Remove the client with age=123 as this is biologically impossible and clearly an erroneous record

In [ ]:
# Remove the outlier client using their row index detected above
# IMPORTANT: Replace the number below with the actual index shown in the cell above output
data.drop(data.index[[39792]], inplace=True)
# Confirm removal by checking the new maximum age value
print("Max age after removal:", data['age'].max())
# Confirm new dataset shape
print("New dataset shape:", data.shape)

Max age after removal: 84.0
New dataset shape: (58644, 12)


Purpose: Detect outliers in max_allowed_loan — negative loan amounts are financially impossible

In [ ]:
# Call IQR function on max_allowed_loan to detect outliers
outliers_loan = find_outliers_IQR(data['max_allowed_loan'])
# Print count of outliers detected
print("number of outliers: " + str(len(outliers_loan)))
# Display records with negative max_allowed_loan specifically
print(data[data['max_allowed_loan'] < 0])

number of outliers: 2233
    age  income home_ownership  emplyment_length loan_intent  loan_amount  loan_interest_rate  loan_income_ratio payment_default_on_file  \
0  21.0   12000            OWN                 0   EDUCATION        15000                6.99               0.12                       N   
1  21.0   13200            OWN                 2   EDUCATION        25000               16.77               0.19                       Y   
2  23.0    9600           RENT                 5     MEDICAL        30000               12.42               0.31                       N   

   credit_history_length  loan_approval_status  max_allowed_loan  
0                      4                     0          -2426900  
1                      3                     0           -111739  
2                      3                     0            -89000  


Purpose: Remove clients with negative maximum loan amounts as these are impossible values

In [ ]:
# Remove all records where max_allowed_loan is negative (impossible value)
data.drop(data[data['max_allowed_loan'] < 0].index, inplace=True)
# Confirm removal — minimum should now be 0 or positive
print("Min max_allowed_loan after removal:", data['max_allowed_loan'].min())
# Confirm new dataset shape
print("New dataset shape:", data.shape)

Min max_allowed_loan after removal: 0
New dataset shape: (58641, 12)


Purpose: Fill missing values in loan_interest_rate with mean — appropriate as only 0.02% missing (under 5% threshold) and variable is continuous numeric

In [ ]:
# Calculate mean of loan_interest_rate for use in imputation
Mean_loan_interest_rate = data['loan_interest_rate'].mean()
# Fill missing values in loan_interest_rate with the calculated mean
data['loan_interest_rate'].fillna(Mean_loan_interest_rate, inplace=True)
# Confirm zero missing values remain in loan_interest_rate
print("Missing in loan_interest_rate after fix:", data['loan_interest_rate'].isnull().sum())

Missing in loan_interest_rate after fix: 0


/tmp/ipykernel_4777/2837382852.py:4: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





Purpose: Fill missing values in age with mean — appropriate as only 0.01% missing and variable is continuous numeric

In [ ]:
# Calculate mean of age variable for imputation
Mean_age = data['age'].mean()
# Fill missing age values with the calculated mean value
data['age'].fillna(Mean_age, inplace=True)
# Confirm zero missing values remain in age
print("Missing in age after fix:", data['age'].isnull().sum())

Missing in age after fix: 0


/tmp/ipykernel_4777/3732583037.py:4: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





Purpose: Remove the 5 records with missing payment_default_on_file — only 0.009% of data so deletion has negligible impact on the dataset

In [ ]:
# Drop all rows that still have any missing values (only 5 records in payment_default_on_file)
data = data.dropna()
# Confirm zero missing values remain across all variables
print("Total missing values remaining:", data.isnull().sum().sum())
# Confirm final dataset shape after all cleaning
print("Dataset shape after all missing value treatment:", data.shape)

Total missing values remaining: 0
Dataset shape after all missing value treatment: (58636, 12)


Purpose: Final verification that all missing values have been resolved

In [ ]:
# Final check — confirm zero missing values remain in all variables
data.isnull().sum()

,0
age,0
income,0
home_ownership,0
emplyment_length,0
loan_intent,0
loan_amount,0
loan_interest_rate,0
loan_income_ratio,0
payment_default_on_file,0
credit_history_length,0


Purpose: Convert age and loan_interest_rate from float to integer as decimal precision is not meaningful for these variables

In [ ]:
# Convert age and loan_interest_rate from float to integer data type
data = data.astype({"age": "int", "loan_interest_rate": "int"})
# Confirm data types have been updated
data.dtypes

,0
age,int64
income,int64
home_ownership,object
emplyment_length,int64
loan_intent,object
loan_amount,int64
loan_interest_rate,int64
loan_income_ratio,float64
payment_default_on_file,object
credit_history_length,int64


Purpose: Encode payment_default_on_file from text (Y/N) to numeric (1/0) using map as ML algorithms cannot process string values

In [ ]:
# Encode payment_default_on_file: Y (payment default) = 1, N (no default) = 0
data['payment_default_on_file'] = data['payment_default_on_file'].map({'Y': 1, 'N': 0})
# Confirm encoding was applied correctly
print("payment_default_on_file unique values:", data['payment_default_on_file'].unique())

payment_default_on_file unique values: [0 1]


Purpose: Import sklearn preprocessing package to encode ordinal categorical variables

In [ ]:
# Import sklearn preprocessing module for label encoding
from sklearn import preprocessing
# Create a LabelEncoder instance for encoding ordinal categorical variables
label_encoder = preprocessing.LabelEncoder()

Purpose: Encode home_ownership and loan_intent from text categories to numeric labels so ML algorithms can process them

In [ ]:
# Encode home_ownership categorical variable to numeric labels
data['home_ownership'] = label_encoder.fit_transform(data['home_ownership'])
# Encode loan_intent categorical variable to numeric labels
data['loan_intent'] = label_encoder.fit_transform(data['loan_intent'])
# Confirm encoding applied — show unique numeric values
print("home_ownership encoded:", data['home_ownership'].unique())
print("loan_intent encoded:", data['loan_intent'].unique())

home_ownership encoded: [3 0 2 1]
loan_intent encoded: [1 2 5 0 4 3]


Purpose: Create the classification dataset for Case Study A by dropping max_allowed_loan which is the regression target only

In [ ]:
# Create classification dataset by dropping the regression target variable
data_classification = data.drop(columns=(['max_allowed_loan']))
# Display dimensions of the classification dataset
print("Classification dataset shape:", data_classification.shape)
# Display all column names in the classification dataset
print("Features:", list(data_classification.columns))

Classification dataset shape: (58636, 11)
Features: ['age', 'income', 'home_ownership', 'emplyment_length', 'loan_intent', 'loan_amount', 'loan_interest_rate', 'loan_income_ratio', 'payment_default_on_file', 'credit_history_length', 'loan_approval_status']


Purpose: Filter only loan-approved clients (loan_approval_status=1) for regression modelling of Maximum Loan Amount

In [ ]:
# Filter only approved clients for the regression dataset
data_regression_temp = data[data['loan_approval_status'] == 1].copy()

# Define the target variable for regression
y_regression = data_regression_temp['loan_amount']

# Define features for regression: drop loan_approval_status (classification target),
# max_allowed_loan (problematic column), and loan_amount (now the target)
X_regression = data_regression_temp.drop(columns=['loan_approval_status', 'max_allowed_loan', 'loan_amount'])

# Display dimensions of the regression datasets
print("Regression features (X) shape:", X_regression.shape)
print("Regression target (y) shape:", y_regression.shape)
# Display all column names in the regression features
print("Features for regression:", list(X_regression.columns))

Regression features (X) shape: (8349, 9)
Regression target (y) shape: (8349,)
Features for regression: ['age', 'income', 'home_ownership', 'emplyment_length', 'loan_intent', 'loan_interest_rate', 'loan_income_ratio', 'payment_default_on_file', 'credit_history_length']


Purpose: Assign all numeric feature names to one variable for use in scaling

In [ ]:
# Define list of all numeric feature columns for scaling (excluding the target 'loan_amount')
numeric_cols = ['age', 'income', 'emplyment_length',
                'loan_interest_rate', 'loan_income_ratio', 'credit_history_length']

Purpose: Assign all categorical feature names to one variable for reference

In [ ]:
# Define list of all categorical feature columns (already encoded to numeric)
categorical_cols = ['home_ownership', 'loan_intent', 'payment_default_on_file']

Purpose: Import MinMaxScaler to scale numeric features to a unified range [0,1]

In [ ]:
# Import MinMaxScaler for normalisation scaling to range [0,1]
from sklearn.preprocessing import MinMaxScaler

Purpose: Scale all numeric features in the classification dataset to [0,1] range to prevent features with larger magnitudes dominating distance-based algorithms such as KNN

In [ ]:
# Create MinMaxScaler instance
scaler = MinMaxScaler()
# Apply Min-Max scaling to numeric columns only in the classification dataset
data_classification[numeric_cols] = scaler.fit_transform(data_classification[numeric_cols])
# Display first 5 rows to confirm scaling was applied
data_classification.head()

,age,income,home_ownership,emplyment_length,loan_intent,loan_amount,loan_interest_rate,loan_income_ratio,payment_default_on_file,credit_history_length,loan_approval_status
3,0.3125,0.093788,3,0.02,1,35000,0.558824,0.228916,0,0.321429,0
4,0.3125,0.045258,0,0.02,2,35000,0.676471,0.469880,0,0.428571,0
5,0.3125,0.066887,0,0.06,5,30000,0.529412,0.277108,0,0.392857,0
6,0.3125,0.076907,0,0.06,0,25000,0.647059,0.204819,0,0.321429,0
7,0.3125,0.067412,0,0.02,2,25000,0.794118,0.265060,0,0.535714,0


Purpose: Scale all numeric features in the regression dataset to [0,1] range

In [ ]:
# Create a new MinMaxScaler instance for the regression features
scaler2 = MinMaxScaler()
# Apply Min-Max scaling to numeric columns in the regression features
X_regression[numeric_cols] = scaler2.fit_transform(X_regression[numeric_cols])
# Display first 5 rows of scaled features to confirm scaling was applied
X_regression.head()

,age,income,home_ownership,emplyment_length,loan_intent,loan_interest_rate,loan_income_ratio,payment_default_on_file,credit_history_length
379,0.000000,0.302421,0,0.01626,0,0.555556,0.219512,0,0.071429
380,0.000000,0.222437,3,0.01626,1,0.500000,0.317073,0,0.071429
381,0.000000,0.127254,3,0.00813,4,0.388889,0.341463,0,0.000000
382,0.016667,0.052550,3,0.04065,2,0.500000,0.256098,0,0.071429
383,0.016667,0.055126,3,0.04065,1,0.333333,0.378049,0,0.000000


Purpose: Save both clean prepared datasets as CSV files for use in Notebooks 2 and 3

In [ ]:
# Save the prepared classification dataset to CSV for use in Notebook 2
data_classification.to_csv('loan_classification_prepared.csv', index=False)

# Combine scaled features and target into a single DataFrame for the regression task
# The 'max_allowed_loan' column was problematic and is now dropped.
# The 'loan_amount' is now the target variable for regression.
data_regression_final = X_regression.copy()
data_regression_final['loan_amount'] = y_regression # Add the unscaled target back

# Save the prepared regression dataset to CSV for use in Notebook 3
data_regression_final.to_csv('loan_regression_prepared.csv', index=False)

# Confirm both files saved successfully
print("Classification dataset saved — shape:", data_classification.shape)
print("Regression dataset saved — shape:", data_regression_final.shape)

Classification dataset saved — shape: (58636, 11)
Regression dataset saved — shape: (8349, 10)
